spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -458,44 +265,73 @@ public Object doWithRetry(RetryContext context) throws RuntimeException {
 				}
 			}
 		}
-		topicsInUse.put(name, listenedPartitions);
-		ReceivingHandler rh = new ReceivingHandler(properties);
-		rh.setOutputChannel(moduleInputChannel);
+		this.topicsInUse.put(name, listenedPartitions);
+		return listenedPartitions;
+	}
 
-		final FixedSubscriberChannel bridge = new FixedSubscriberChannel(rh);
-		bridge.setBeanName("bridge." + name);
 
+	@Override
+	@SuppressWarnings("unchecked")
+	protected MessageProducer createConsumerEndpoint(String name, String group, Object queue,
+			ExtendedConsumerProperties<KafkaConsumerProperties> properties) {
+
+		Collection<Partition> listenedPartitions = (Collection<Partition>) queue;
 		Assert.isTrue(!CollectionUtils.isEmpty(listenedPartitions), "A list of partitions must be provided");
+
+		int concurrency = Math.min(properties.getConcurrency(), listenedPartitions.size());
+
+		final ExecutorService dispatcherTaskExecutor =
+				Executors.newFixedThreadPool(concurrency, DAEMON_THREAD_FACTORY);
 		final KafkaMessageListenerContainer messageListenerContainer = new KafkaMessageListenerContainer(
-				connectionFactory, listenedPartitions.toArray(new Partition[listenedPartitions.size()]));
+				this.connectionFactory, listenedPartitions.toArray(new Partition[listenedPartitions.size()])) {
 
-		if (logger.isDebugEnabled()) {
-			logger.debug("Listened partitions: " + StringUtils.collectionToCommaDelimitedString(listenedPartitions));
+			@Override
+			public void stop(Runnable callback) {
+				super.stop(callback);
```

Comment:
If we need to ensure the superclass has closed, we should be passing a new callback to it (and waiting for it to be invoked), not the one that is passed here as a parameter (which is used by the application context when closing). The "callback" param should instead be invoked at the end of *this* method, so that the context does not think it's completed prematurely.